File: apps/web/lib/report.ts
import type { CgmReportResponse } from "@/lib/types/api";

/**
 * Pure shaping helpers for the clinical report page. No React, no fetch — kept
 * here so the formatting and the band-segment math are unit-testable in
 * isolation. The numeric formulas all live in Python (core.metrics); this file
 * only formats and arranges already-computed values for display.
 */

/** Band palette — mirrors apps/local/charts/report.py (BG day-view palette). */
export const bandColors = {
  tbr2: "#B71C1C", // very low
  tbr1: "#E53935", // low
  tir: "#2E7D32", // in range
  tar1: "#FB8C00", // high
  tar2: "#E65100", // very high
} as const;

export interface BandSegment {
  key: "tbr2" | "tbr1" | "tir" | "tar1" | "tar2";
  label: string;
  pct: number;
  color: string;
}

const BAND_DEFS: ReadonlyArray<Pick<BandSegment, "key" | "label">> = [
  { key: "tbr2", label: "Very low" },
  { key: "tbr1", label: "Low" },
  { key: "tir", label: "In range" },
  { key: "tar1", label: "High" },
  { key: "tar2", label: "Very high" },
];

/**
 * Build the five stacked time-in-bands segments (tbr2→tar2) from a report.
 * Negative or non-finite values are clamped to 0 so the bar never renders a
 * bogus segment; the order matches the clinical low→high convention.
 */
export function buildBandSegments(report: CgmReportResponse): BandSegment[] {
  return BAND_DEFS.map(({ key, label }) => {
    const raw = report[key];
    const pct = Number.isFinite(raw) && raw > 0 ? raw : 0;
    return { key, label, pct, color: bandColors[key] };
  });
}

/**
 * Format a nullable metric for a tile. `null` → the em-dash placeholder.
 * `digits` controls decimal places; `suffix` is appended (e.g. "%", " mg/dL").
 */
export function formatMetric(
  value: number | null | undefined,
  opts: { digits?: number; suffix?: string } = {},
): string {
  if (value === null || value === undefined || !Number.isFinite(value)) {
    return "—";
  }
  const { digits = 1, suffix = "" } = opts;
  return `${value.toFixed(digits)}${suffix}`;
}

/** Sufficiency note shown when the window is below the consensus gate. */
export function sufficiencyNote(report: CgmReportResponse): string | null {
  if (report.meets_sufficiency) return null;
  return (
    `Data sufficiency not met (need ≥14 days and ≥70% active CGM time; ` +
    `have ${report.days_covered} days, ${report.active_pct.toFixed(0)}% active). ` +
    `GMI and GRI are withheld until the window is sufficient.`
  );
}

File: apps/local/charts/report.py
"""Clinical report visuals — the time-in-bands stacked bar.

Pure chart builder (no Streamlit import) so it stays unit-testable. The page
itself (apps/local/app.py) renders the scalar tiles via ``st.metric`` from a
``core.metrics.CgmReport``.
"""

from __future__ import annotations

import plotly.graph_objects as go

# Band colors — align with the day-view BG palette (low→red, in-range→green,
# high→orange, very-high→deep orange).
_BANDS = [
    ("tbr2", "Very low", "#B71C1C"),
    ("tbr1", "Low", "#E53935"),
    ("tir", "In range", "#2E7D32"),
    ("tar1", "High", "#FB8C00"),
    ("tar2", "Very high", "#E65100"),
]


def build_time_in_bands_bar(
    *,
    tbr2: float,
    tbr1: float,
    tir: float,
    tar1: float,
    tar2: float,
) -> go.Figure:
    """Single horizontal stacked bar of the five consensus glucose bands."""
    values = {"tbr2": tbr2, "tbr1": tbr1, "tir": tir, "tar1": tar1, "tar2": tar2}
    fig = go.Figure()
    for key, label, color in _BANDS:
        pct = float(values[key])
        fig.add_trace(
            go.Bar(
                x=[pct],
                y=["Time in bands"],
                name=label,
                orientation="h",
                marker_color=color,
                hovertemplate=f"{label}: <b>{pct:.1f}%</b><extra></extra>",
            )
        )
    fig.update_layout(
        barmode="stack",
        height=120,
        margin=dict(l=10, r=10, t=10, b=10),
        showlegend=True,
        legend=dict(orientation="h", yanchor="bottom", y=1.05, x=0, font=dict(size=10)),
        xaxis=dict(range=[0, 100], ticksuffix="%", showgrid=False),
        yaxis=dict(showticklabels=False),
        plot_bgcolor="white",
        paper_bgcolor="white",
    )
    return fig
